proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,477 @@
+  otherwise throw `InvalidArgument`
+  - Set the length of the program account to 0 (removing the header too)
+  - Transfer all lamports from the program account to the recipient account
+
+#### Deploy
+
+- Instruction accounts:
+  - `[writable]` The program account to deploy.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `5u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account header
+  - Check that the status stored in the program account is `NeverBeenDeployed`
+  or `Retracted`
+    otherwise throw `InvalidArgument`
+  - If the status is `Retracted` then also check that the slot stored in the
+  program account is not the current (deployment cooldown),
+  otherwise throw `InvalidArgument`
+  - Charge program_length_in_bytes / cpi_bytes_per_unit CUs
+  - Check that the executable file stored in the program account passes
+  executable verification
+  - Change the slot in the program account to the current slot
+  - Change the status stored in the program account to `Deployed`
+  - Set the `is_executable` flag to `true`
+
+#### Retract
+
+- Instruction accounts:
+  - `[writable]` The program account to retract.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `6u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account header
+  - Check that the status stored in the program account is `Deployed`,
+    otherwise throw `InvalidArgument`
+  - Check that the slot stored in the program account is not the current
+    (deployment cooldown), otherwise throw `InvalidArgument`
+  - Note: The slot is **not** set to the current slot to allow a
+  retract-modify-redeploy-sequence within the same slot or even within the
+  same transaction.
+  - Change the status stored in the program account to `Retracted`
+  - Set the `is_executable` flag to `false`
+
+#### Initialize
+
+- Instruction accounts:
+  - `[writable]` The account to initialize as program account.
+  - `[signer]` The new authority of the program.
+- Instruction data:
+  - Enum variant `7u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Check that the owner of the program account is loader-v4,
+    otherwise throw `InvalidAccountOwner`
+  - Check that the program account is writable,
+    otherwise throw `InvalidArgument`
+  - Check that there are enough funds in the program account for rent
+  exemption of the header,
+    otherwise throw `InsufficientFunds`
+  - Check that the new authority (instruction account at index 1) has signed,
+    otherwise throw `MissingRequiredSignature`
+  - Set the length of the program account to the header size
+  - Set the slot to zero, **not** the current slot
```

Comment:
Why is this important? I don't think it matters what the slot is set to for accounts with `NeverBeenDeployed` status right?